src/main/data/BigBed.js

```diff
@@ -30,21 +30,59 @@ function parseCirTree(buffer) {
   return new jBinary(buffer, bbi.TYPE_SET).read('CirTree');
 }
 
-// Extract a map from contig name --> contig ID from the bigBed header.
-function generateContigMap(header): {[key:string]: number} {
-  // Just assume it's a flat "tree" for now.
-  var nodes = header.chromosomeTree.nodes.contents;
+//
+//Extract a map from contig name --> contig ID from the bigBed header.
+//
+function generateContigMap(header, remoteFile): Q.Promise<{ [key: string]: number }> {
+  return generateContigMapNodes(header.chromosomeTree.nodes, header.chromosomeTree.keySize, remoteFile);
+}
+
+//Transforms nodes into contig name --> contig ID
+function generateContigMapNodes(nodes, keySize, remoteFile): Q.Promise<{ [key: string]: number }> {
+  if (nodes.isLeaf) {
+    return generateContigMapFromLeafNode(nodes.contents);
+  } else {
+    return generateContigMapFromNonLeafNode(nodes.contents, keySize, remoteFile);
+  }
+}
+
+//Transforms leaf nodes into contig name --> contig ID
+function generateContigMapFromLeafNode(nodes): Q.Promise<{ [key: string]: number }> {
   if (!nodes) {
     throw 'Invalid chromosome tree';
   }
-  return _.object(nodes.map(function({id, key}) {
+  return Q.resolve(_.object(nodes.map(function ({id, key}) {
     // remove trailing nulls from the key string
     return [key.replace(/\0.*/, ''), id];
-  }));
+  })));
+}
+
+//Transforms non leaf nodes into contig name --> contig ID. It requires fetching data for all children nodes.
+function generateContigMapFromNonLeafNode(nodes, keySize, remoteFile): Q.Promise<{ [key: string]: number }> {
+  return Q.all(nodes.map(function ({key, offset}) {
+    return remoteFile.getBytes(offset, 64 * 1024).then(buffer => {
+      return parseNodeLeaf(buffer, keySize);
+    }).then(childNodeData => generateContigMapNodes(childNodeData, keySize, remoteFile));
+  })).then(data => {
+    return _.reduce(data, function (memo, item) {
+      return _.extend(memo, item);
+    }, {});
+  });
+}
+
+//
+// Parse Contig Tree Node assuming that we know the keySize. This method is used for recursive parsing of the tree.
+//
+function parseNodeLeaf(buffer: ArrayBuffer, keySize: number) {
+  //use a copy of the global definition and replace variable with constant (this variable would be unavailable here)
+  var typeCopy = _.extend({}, bbi.TYPE_SET);
+  typeCopy['BPlusTreeNode']['contents'][1][2]['key'][1] = keySize;
+  typeCopy['BPlusTreeNode']['contents'][1][3]['key'][1] = keySize;
+  return new jBinary(buffer, typeCopy).read('BPlusTreeNode');
 }
 
 // Generate the reverse map from contig ID --> contig name.
-function reverseContigMap(contigMap: {[key:string]: number}): Array<string> {
+function reverseContigMap(contigMap: { [key: string]: number }): Array<string> {
```

Comment:
It looks like your editor modifier a lot of the spacing/indentation in this file. Would you be able to revert the lines you did not intend to modify?

Reply:
Sorry for that.
I reverted the indentation